Sprint-2/Step-1

Ekaterina64 opened the pull request:
Добавлено:
- Контекст данных ингредиентов
- Контекст итоговой стоимости
- Расчет стоимости заказа
- Обработка запроса на заказ
- Передача номера заказа в модальное окно

Ekaterina-ux committed:
Добавлено:
- Контекст данных ингредиентов
- Контекст итоговой стоимости
- Расчет стоимости заказа
- Обработка запроса на заказ
- Передача номера заказа в модальное окно

gennady-bars commented on src/utils/api.js:
```diff
@@ -0,0 +1,17 @@
+
+export function getIngredients() {
+	return fetch(`${NORMA_API}/ingredients`)
+	 .then(checkResponse)
```
**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -60,31 +35,13 @@ const App = () => {
-        <>
+        <IngredientsDataContext.Provider value={state.data}>
           <AppHeader/>
           <div className={classNames(styles.burgerContainer, "pl-5 pr-5")}>
```
Тут `div` нужно заменить на `main`

По семантике нужно помещать весь главный контент страницы после хэдера в тег `main`.

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -60,31 +35,13 @@ const App = () => {
     <>
       { state.data &&
-        <>
+        <IngredientsDataContext.Provider value={state.data}>
```
Хороший контекст сделали

gennady-bars commented on src/components/burger-constructor/burger/filling-list/filling-item/filling_item.jsx:
```diff
@@ -3,7 +3,7 @@ import {
 import PropTypes from "prop-types"
-import styles from "../../burger-constructor.module.css"
+import styles from "../../../burger-constructor.module.css"
 
```
`filling_item.jsx` нужно сделать в `kebab-case`, как другие файлы по заданию

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -1,14 +1,29 @@
+	const data = useContext(IngredientsDataContext);
 	const [ dataBun, dataMain, dataSouse ] = useFilter(data);
+
+	const [isModalIngredientsDetailsOpened, setIsModalIngredientsDetailsOpened] = useState(false);
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для управления модальным окном:

```js
import { useState, useCallback } from "react";

// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export const useModal = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  // `useCallback` нужен для того, чтобы зафиксировать ссылку на функцию. Таким образом уменьшится кол-во перерисовок компонента, куда будет передана эта функция
  const openModal = useCallback(() => {
    setIsModalOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsModalOpen(false);
  }, []);

  return {
    isModalOpen,
    openModal,
    closeModal,
  };
};

```

Этот код помещают в отдельный файл `useModal.js` в папке `hooks` и импортируют функцию туда, где нужно создать показатели открытости/закрытости попапа (у каждого попапа они свои)

И теперь нужно просто вызывать хук в том компоненте, где нужно управлять попапом. Всего одна строчка кода будет:

```js
  const { isModalOpen, openModal, closeModal } = useModal();

````

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Хороший контекст сделали



но есть некоторые недочеты:



* Скрин https://disk.yandex.ru/i/yPLJ2XQb2J31lw видна граница оверлея . Модальное окно должно быть всегда ровно посередине экрана. Для этого нужно указать ему и оверлею в `css` позицию `fixed`.
* Скрин https://disk.yandex.ru/i/v4F39OJJP7oiuw хэдер не должен сужаться, если под ним контент в нормальном состоянии
* По семантике нужно помещать весь главный контент страницы после хэдера в тег `main`.
* `filling_item.jsx` нужно сделать в `kebab-case`, как другие файлы по заданию

Ekaterina-ux committed:
Исправлено:
- Отображение оверлея
- Верстка и стили хедера
- Главный контент страницы в main
- Названия файлов по kebab-case
Улучшено:
- Универсальная функция запроса с проверкой ответа
- Добавлен кастомный хук useModal

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
